text/3778-inline-intents.md

```diff
@@ -0,0 +1,189 @@
+- Feature Name: `inline_intents`
+- Start Date: 2025-02-22
+- RFC PR: [rust-lang/rfcs#3778](https://github.com/rust-lang/rfcs/pull/3778)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add `#[inline(trampoline)]` and `#[inline(rarely)]` to give additional control
+over the inlining decisions made by the compiler.
+
+# Motivation
+[motivation]: #motivation
+
+Right now it's pretty common to just slap `#[inline]` on things without thinking
+about it all that hard about it, and the existing controls aren't great.
+
+Often we'll get PRs using `inline(always)` "because it's just calling something
+else so of course it should be inlined", for example.  But because of the
+bottom-up nature of inlining, that's a bad thing to do because if the callee
+happens to get inlined, then it'll "always" inline that callee too, which might
+not be what was actually desired.
+
+At the same time, sometimes it's useful to put `inline` on things to make the
+definition available to to LLVM, but where it probably shouldn't actually be
+inlined in general, only in particular special cases (perhaps when one of the
+arguments is a small constant, for example).
+
+It would thus be nice to give additional options to the user to let them
+describe *why* they wanted a function inlined, and hopefully be able to make
+better decisions in the backend as a result.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In most cases, plain `#[inline]` is fine, especially with PGO and LTO.
+However, if you've measured and things are making a poor choice, there are some
+options you can use to hint the compiler towards what you want.
+
+## `inline(trampoline)`
+
+This is intended for functions which quickly "bounce" the caller off to some
+other implementation, after doing some initial checks or transformations.
```

Comment:
I think this could usefully be a little more explicit:

```suggestion
This is intended for functions which quickly "bounce" the caller off to some
other implementation which should not necessarily be inlined itself,
after doing some initial checks or transformations.
```